src/python/pants/backend/jvm/tasks/junit_run.py

```diff
@@ -575,13 +585,65 @@ def files_iter():
           yield os.path.join(dir_path, filename)
     return list(files_iter())
 
-  def _execute(self, all_targets):
-    # NB: We only run tests within junit_tests targets, but if coverage options are
-    # specified, we want to instrument and report on all the original targets, not
-    # just the test targets.
-    partition = all_targets if self.get_options().coverage else self._get_test_targets()
+  def _iter_partitions(self, targets, output_dir):
+    if self._per_target:
+      for target in targets:
+        yield (target,), os.path.join(output_dir, target.id)
+    else:
+      yield tuple(targets), output_dir
 
-    with self.invalidated(targets=partition,
+  def _execute(self, all_targets):
+    with self._isolation(all_targets) as (output_dir, reports, coverage):
+      results = {}
+      failure = False
+      for (partition, partition_output_dir) in self._iter_partitions(self._get_test_targets(),
+                                                                     output_dir):
+        try:
+          rv = self._run_partition(partition, partition_output_dir, coverage)
+        except ErrorWhileTesting as e:
+          rv = TestResult.from_error(e)
+
+        results[partition] = rv
+        if not rv.success:
+          failure = True
+          if self._fail_fast:
+            break
+
+      for partition in sorted(results):
```

Comment:
Should we pre-sort the partitions for stability and then use an OrderedDict for results to preserve the sorting?

Reply:
I think not. This is already done super indirectly via a root OrderedDict in BuildGraph that feeds Context ... etc. IE: The local code here steps through partitions in context.targets() order, preserving whatever order that imposes, which presumably is the order that makes sense if any does.

FWIW, if further discussion highlights this should be done, PytestRun needs the fix too, leading to your next comment below.